### PR TITLE
Adds primitive implementations of DISTINCT

### PR DESCRIPTION
**Pull request by systay**
The general purpose DISTINCT implementation uses a
lot of objects to make the output distinct.
This implementation avoids objects and uses primitive
sets to do work, which should be create less memory
pressure and is easier for the CPU to do efficiently.

Replaces #11627

### CODE REVIEW

File: community/cypher/runtime-util/src/main/java/org/neo4j/cypher/internal/runtime/LongArraySet.java [hunk 1]
```diff
@@ -0,0 +1,334 @@
...
+                    }
+                }
+            }
+        }
+    }
+
+    /***
+     * Returns true if the value is in the set.
+     * @param value The value to check for
+    1     * @return whether the value is in the set or not.
```
Comment (pontusmelke) [inline community/cypher/runtime-util/src/main/java/org/neo4j/cypher/internal/runtime/LongArraySet.java]: `1`?

Comment (systay) [reply community/cypher/runtime-util/src/main/java/org/neo4j/cypher/internal/runtime/LongArraySet.java]: Remind me not to program with buttered fingers again, please.

File: community/cypher/runtime-util/src/main/scala/org/neo4j/cypher/internal/runtime/PrefetchingIterator.scala [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+  def produceNext(): Option[T]
+
+  override def hasNext: Boolean = {
+    if (buffer == null)
+      pullNextElementFromSource()
+    buffer.nonEmpty
+  }
+
+  override def next(): T = {
+    if (buffer == null)
```
Comment (pontusmelke) [inline community/cypher/runtime-util/src/main/scala/org/neo4j/cypher/internal/runtime/PrefetchingIterator.scala]:
```
if (!hasNext) throw NoSuchElementException

val current = buffer.get
pullNextElementFromSource()
current
```
?

File: community/cypher/runtime-util/src/main/java/org/neo4j/cypher/internal/runtime/LongArraySet.java [hunk 3]
```diff
@@ -0,0 +1,334 @@
...
+
+    /**
+     * Adds a value to the set.
+     *
+     * @param value The new value to be added to the set
+     * @return The method returns true if the value was added and false if it already existed in the set.
+     */
+    public boolean add( long[] value )
+    {
+        assert value.length == width : "all elements must have the same size";
```
Comment (fickludd) [inline community/cypher/runtime-util/src/main/java/org/neo4j/cypher/internal/runtime/LongArraySet.java]: How about adding another `assert` here, checking that `value` does not contain unsupported values (e.g. `< 0` or `== 0xF000000000000000L`)?

File: community/cypher/runtime-util/src/main/java/org/neo4j/cypher/internal/runtime/LongArraySet.java [hunk 4]
```diff
@@ -0,0 +1,334 @@
...
+    }
+
+    /***
+     * Returns true if the value is in the set.
+     * @param value The value to check for
+    * @return whether the value is in the set or not.
+     */
+    public boolean contains( long[] value )
+    {
+        assert value.length == width : "all elements must have the same size";
```
Comment (fickludd) [inline community/cypher/runtime-util/src/main/java/org/neo4j/cypher/internal/runtime/LongArraySet.java]: Could assert that `value` is valid here too.

File: community/cypher/runtime-util/src/main/java/org/neo4j/cypher/internal/runtime/LongArraySet.java [hunk 5]
```diff
@@ -0,0 +1,334 @@
...
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.neo4j.cypher.internal.runtime;
+
+/**
+ * When you need to have a set of arrays of longs representing entities - look no further
+ * <p>
+ * This set will keep all it's state in a single long[] array, marking unused slots
+ * using 0xF000000000000000L, a value that should never be used for node or relationship id's.
```
Comment (fickludd) [inline community/cypher/runtime-util/src/main/java/org/neo4j/cypher/internal/runtime/LongArraySet.java]: This was changed to `-2` right?

File: enterprise/cypher/slotted-runtime/src/main/scala/org/neo4j/cypher/internal/runtime/slotted/pipes/DistinctSlottedPrimitivePipe.scala [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+
+  projections.values.foreach(_.registerOwningPipe(this))
+
+  //===========================================================================
+  // Runtime code
+  //===========================================================================
+  protected def internalCreateResults(input: Iterator[ExecutionContext],
+                                      state: QueryState): Iterator[ExecutionContext] = {
+    new PrefetchingIterator[ExecutionContext] {
+      private val seen = new LongArraySet(32, projections.size)
```
Comment (fickludd) [inline enterprise/cypher/slotted-runtime/src/main/scala/org/neo4j/cypher/internal/runtime/slotted/pipes/DistinctSlottedPrimitivePipe.scala:58]: Is this the same as `new LongArraySet(32, primitiveSlots.length)`?

**Review by fickludd — APPROVED** [review]
I have some small suggestions to fix or forget, but otherwise looks good!

File: enterprise/cypher/slotted-runtime/src/main/scala/org/neo4j/cypher/internal/runtime/slotted/pipes/DistinctSlottedPrimitivePipe.scala (hunk 6 above)
Comment (systay) [reply enterprise/cypher/slotted-runtime/src/main/scala/org/neo4j/cypher/internal/runtime/slotted/pipes/DistinctSlottedPrimitivePipe.scala:58]: Yep

**Pull request merged**